seurat/CellClassificationFXs.R

```diff
@@ -0,0 +1,605 @@
+      genes.list[[i]] <- names(x = which(x = object@kmeans.obj[[1]]$cluster == i))
+    }
+    cluster.length <- length(x = genes.list)
+  } else {
+    if (is.null(x = genes.list)) {
+      stop("Missing input gene list")
+    }
+    genes.list <- lapply(
+      X = genes.list,
+      FUN = function(x) {
+        return(intersect(x = x, y = rownames(object)))
+      }
+    )
+    cluster.length <- length(x = genes.list)
+  }
+  if (!all(LengthCheck(values = genes.list))) {
+    warning(paste(
+      'Could not find enough genes in the object from the following gene lists:',
+      paste(names(x = which(x = ! LengthCheck(values = genes.list)))),
+      'Attempting to match case...'
+    ))
+    genes.list <- lapply(
+      X = genes.old,
+      FUN = CaseMatch, match = rownames(object)
+    )
+  }
+  if (!all(LengthCheck(values = genes.list))) {
+    stop(paste(
+      'The following gene lists do not have enough genes present in the object:',
+      paste(names(x = which(x = ! LengthCheck(values = genes.list)))),
+      'exiting...'
+    ))
+  }
+  if (is.null(x = genes.pool)) {
+    genes.pool = rownames(object)
+  }
+  data.avg <- Matrix::rowMeans(object@assays$RNA@data[genes.pool, ])
+  data.avg <- data.avg[order(data.avg)]
+  data.cut <- as.numeric(x = Hmisc::cut2(
+    x = data.avg,
+    m = round(x = length(x = data.avg) / n.bin)
+  ))
+  names(x = data.cut) <- names(x = data.avg)
+  ctrl.use <- vector(mode = "list", length = cluster.length)
+  for (i in 1:cluster.length) {
+    genes.use <- genes.list[[i]]
+    for (j in 1:length(x = genes.use)) {
+      ctrl.use[[i]] <- c(
+        ctrl.use[[i]],
+        names(x = sample(
+          x = data.cut[which(x = data.cut == data.cut[genes.use[j]])],
+          size = ctrl.size,
+          replace = FALSE
+        ))
+      )
+    }
+  }
+
+  ctrl.use <- lapply(X = ctrl.use, FUN = unique)
+  ctrl.scores <- matrix(
+    data = numeric(length = 1L),
+    nrow = length(x = ctrl.use),
+    ncol = ncol(x = object@assays$RNA@data)
+  )
+
+  for (i in 1:length(ctrl.use)) {
+    genes.use <- ctrl.use[[i]]
+    ctrl.scores[i, ] <- Matrix::colMeans(x = object@assays$RNA@data[genes.use, ])
+  }
+  genes.scores <- matrix(
+    data = numeric(length = 1L),
+    nrow = cluster.length,
+    ncol = ncol(x = object@assays$RNA@data)
+  )
+  for (i in 1:cluster.length) {
+    genes.use <- genes.list[[i]]
+    data.use <- object@assays$RNA@data[genes.use, , drop = FALSE]
+    genes.scores[i, ] <- Matrix::colMeans(x = data.use)
+  }
+  genes.scores.use <- genes.scores - ctrl.scores
+  rownames(x = genes.scores.use) <- paste0(enrich.name, 1:cluster.length)
+  genes.scores.use <- as.data.frame(x = t(x = genes.scores.use))
+  rownames(x = genes.scores.use) <- colnames(x = object@assays$RNA@data)
+
+  if(!returnScore){
+
+    for (colName in colnames(genes.scores.use)) {
+      object[[colName]] <- genes.scores.use[colnames(object),colName]
+    }
+
+    gc(verbose = FALSE)
+    return(object)
+
+  } else {
```

Comment:
Same as my comment below.  in general, if you have an if/else, and one of these conditions is a few lines of code that stops/exits the method (like the else here), it tends to be cleaner to put that upfront, and then you dont need to wrap the longer if() portion